xds/csds/csds.go

```diff
@@ -0,0 +1,302 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package csds implements features to dump the status (xDS responses) the
+// xds_client is using.
+//
+// Notice: This package is EXPERIMENTAL and may be changed or removed in a later
+// release.
+package csds
+
+import (
+	"context"
+	"io"
+	"time"
+
+	v3adminpb "github.com/envoyproxy/go-control-plane/envoy/admin/v3"
+	v2corepb "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3statuspb "github.com/envoyproxy/go-control-plane/envoy/service/status/v3"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/xds/internal/client"
+	"google.golang.org/grpc/xds/internal/client/bootstrap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	_ "google.golang.org/grpc/xds/internal/client/v2" // Register v2 xds_client.
+	_ "google.golang.org/grpc/xds/internal/client/v3" // Register v3 xds_client.
+)
+
+// xdsClientInterface contains methods from xdsClient.Client which are used by
+// the server. This is useful for overriding in unit tests.
+type xdsClientInterface interface {
+	WatchListener(string, func(client.ListenerUpdate, error)) func()
+	WatchRouteConfig(string, func(client.RouteConfigUpdate, error)) func()
+	WatchCluster(string, func(client.ClusterUpdate, error)) func()
+	WatchEndpoints(string, func(client.EndpointsUpdate, error)) func()
+	DumpLDS() (string, map[string]client.UpdateWithMD)
+	DumpRDS() (string, map[string]client.UpdateWithMD)
+	DumpCDS() (string, map[string]client.UpdateWithMD)
+	DumpEDS() (string, map[string]client.UpdateWithMD)
+	BootstrapConfig() *bootstrap.Config
+	Close()
+}
+
+var (
+	logger       = grpclog.Component("xds")
+	newXDSClient = func() (xdsClientInterface, error) {
+		return client.New()
+	}
+)
+
+type clientStatusServer struct {
+	// xdsClient will always be the same in practise. But we keep a copy in each
+	// server instance for testing.
+	xdsClient    xdsClientInterface
+	xdsClientErr error
+}
+
+// NewClientStatusDiscoveryServer returns an implementation of the CSDS server that can be
+// registered on a gRPC server.
+func NewClientStatusDiscoveryServer() v3statuspb.ClientStatusDiscoveryServiceServer {
+	ret := &clientStatusServer{}
+	ret.xdsClient, ret.xdsClientErr = newXDSClient()
+	if ret.xdsClientErr != nil {
+		logger.Errorf("failed to create xds client: %v", ret.xdsClientErr)
```

Comment:
I'm not convinced by the argument that returning an error makes the function harder to use. And that if the client creation runs into an error here, it would have elsewhere as well.
Given that this is a public facing API, I feel it would be better to have it stand on its own rather than have these unsaid dependencies.

Reply:
I changed this to return an error.

Also the returned struct is a concrete struct, instead of the interface. So we can also get rid of the finalizer.